optimize-nextflow.py

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/env python3
+
+import json
+import shutil
+from collections import namedtuple
+from dataclasses import astuple, dataclass
+from functools import cached_property
+from math import ceil
+from pathlib import Path
+from textwrap import dedent, indent
+from typing import Generic, TypeVar
+
+import typer
+
+Number = TypeVar("Number", int, float)
+
+app = typer.Typer(
+    context_settings={"max_content_width": shutil.get_terminal_size().columns}
+)
+
+
+@app.command()
+def from_json(metrics_path: Path):
+    """Process the metrics JSON output from the Tower CLI.
+
+    Here's an example Tower CLI command:
+
+        tw --output "json" runs view -w "<org>/<workspace>" -i "<workflow-id>" metrics
+    """
+    json_parser = MetricsJsonParser(metrics_path)
+    task_metrics = json_parser.parse()
+    config_generator = NextflowConfigGenerator(task_metrics)
+    config = config_generator.generate()
+    print(config)
+
+
+@app.command()
+def from_api():
+    """Process the metrics JSON output from the Tower API.
+
+    Not implemented yet.
+    """
+    ...
+
+
+@dataclass
+class Stats(Generic[Number]):
+    q1: Number
+    q2: Number
+    q3: Number
+    min: Number
+    max: Number
+    mean: Number
+
+    @cached_property
+    def max_q3_ratio(self):
+        return round(self.max / self.q3, 2)
+
+
+@dataclass
+class TaskMetrics:
+    name: str
+    mem: Stats
+    mem_pct: Stats
+    cpu: Stats
+    cpu_pct: Stats
+
+    @staticmethod
+    def bytes_to_gb(memory: int) -> int:
+        memory_gb = memory / 1024**3
+        return ceil(memory_gb + 0.5)
+
+    @staticmethod
+    def load_to_cores(load: float) -> int:
+        cores = load / 100
+        return ceil(cores + 0.5)
```

Comment:
What is your rationate for adding 0.5 here? (same question for GB).

Reply:
It's for adding a buffer. If the number of CPU cores is close to the ceiling (_i.e._ above X.5), by adding 0.5 ensures that the ceiling will be the next integer. For example, 366% CPU load would translate to 4 raw CPU cores, but I want 5 CPU cores so there's some buffer. I added docstrings to mention this buffer.